perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/ConsolidatorManagerImpl.java

```diff
@@ -460,6 +458,184 @@ public List<UserExtSource> consolidateIdentityUsingToken(PerunSession sess, Stri
+					StringUtils.join(Arrays.asList(ues.getLogin(), ues.getExtSource().getName(), ues.getExtSource().getType()), " | "));
+				throw ex;
+			} else {
+				IdentityAlreadyInUseException ex = new IdentityAlreadyInUseException("Your identity is already associated with a different user. If you are really the same person, please contact support to help you.", user1, user2);
+				log.warn("Identity to be joined is already used by different user. Current user: {}, Current identity: {}, Original user: {}, Original identity: {}",
+					user1, StringUtils.join(Arrays.asList(sess.getPerunPrincipal().getActor(), sess.getPerunPrincipal().getExtSourceName(), sess.getPerunPrincipal().getExtSourceType()), " | "),
+					user2, StringUtils.join(Arrays.asList(ues.getLogin(), ues.getExtSource().getName(), ues.getExtSource().getType()), " | "));
+				throw ex;
+			}
+		}
+		UserExtSource createdUes;
+		if(user1 != null) {
+			createdUes = createExtSourceAndUserExtSource(user1, userInfo2.sub,
+				userInfo2.issuer, sess.getPerunPrincipal().getExtSourceType(),
+				sess.getPerunPrincipal().getExtSourceLoa());
+			Map<String, String> additionalInformations = new HashMap<>();
+			getAdditionalInformations(accessToken, additionalInformations);
+
+			((PerunBlImpl)perun).setUserExtSourceAttributes(sess, createdUes, additionalInformations);
+			log.info("{} joined identities. Current identity: {}, Original identity: {}", user1,
+				StringUtils.join(Arrays.asList(sess.getPerunPrincipal().getActor(), sess.getPerunPrincipal().getExtSourceName(), sess.getPerunPrincipal().getExtSourceType(), " | ")),
+				StringUtils.join(Arrays.asList(userInfo2.sub, userInfo2.issuer, sess.getPerunPrincipal().getExtSourceType()), " | "));
+
+		} else {
+			createdUes = createExtSourceAndUserExtSource(user2, sess.getPerunPrincipal().getActor(),
+				sess.getPerunPrincipal().getExtSourceName(), sess.getPerunPrincipal().getExtSourceType(),
+				sess.getPerunPrincipal().getExtSourceLoa());
+			((PerunBlImpl)perun).setUserExtSourceAttributes(sess, createdUes, sess.getPerunPrincipal().getAdditionalInformations());
+			log.info("{} joined identities. Current identity: {}, Original identity: {}", user2,
+				StringUtils.join(Arrays.asList(sess.getPerunPrincipal().getActor(), sess.getPerunPrincipal().getExtSourceName(), sess.getPerunPrincipal().getExtSourceType(), " | ")),
+				StringUtils.join(Arrays.asList(userInfo2.sub, userInfo2.issuer, sess.getPerunPrincipal().getExtSourceType()), " | "));
+
+		}
+
+		AuthzResolverBlImpl.refreshSession(sess);
+
+		if (BeansUtils.getCoreConfig().isSendIdentityAlerts()) {
+			try {
+				Utils.sendIdentityAddedAlerts(sess, createdUes);
+			} catch (Exception e) {
+				log.error("Failed to send identity added alerts.", e);
+			}
+		}
+		return null;
+	}
+
+	private record OidcUserInfo(String sub, String issuer, Object userInfo) {}
+
+	private static OidcUserInfo getOidcSubIssuer(String accessToken, String issuer) throws PerunException {
```

Comment:
the same code as in the Api class - needs to be refactored so we don't have completely the same code on the two places